File: src/main/java/com/github/tammo/fun/frontend/type/TypeChecker.java
package com.github.tammo.fun.frontend.type;

import com.github.tammo.fun.frontend.ast.SyntaxNode;

public class TypeChecker {

    public static TypedTreeNode typeCheck(SyntaxNode tree) {
        final var annotatedTypedTree = TypeAnnotate.annotateTypes(tree);
        final var constraints = ConstraintsCollector.collect(annotatedTypedTree);

        final var unifier = new Unifier();
        unifier.unifyAll(constraints);

        return new SubstitutionApplier(unifier).apply(annotatedTypedTree);
    }

}

File: src/main/scala/com/github/tammo/frontend/type/TypeChecker.scala
package com.github.tammo.frontend.`type`

import com.github.tammo.diagnostics.CompilerError.TypeCheckError
import com.github.tammo.frontend.ast.SyntaxTree

object TypeChecker {

  def typeCheck(tree: SyntaxTree.CompilationUnit): Either[Seq[TypeCheckError], TypedTree] = {
    val annotatedTypeTree = TypeAnnotate.annotateTypes(tree)
    val constraints = Constraints.collect(annotatedTypeTree)
    Unifier.unifyAll(constraints).map { substitutions =>
      SubstitutionApplier.applySubstitutions(
        annotatedTypeTree,
        `type` => Unifier.applySubstitution(substitutions, `type`)
      )
    }
  }
}
